client/charts/js/components/Autocomplete.jsx

```diff
@@ -0,0 +1,173 @@
+import React, {useReducer} from 'react'
+import classNames from "classnames"
+
+import PropTypes from 'prop-types'
+
+
+const initialState = {
+	// The active selection's index
+    activeSuggestion: 0,
+    // The suggestions that match the user's input
+    filteredSuggestions: [],
+    // Whether or not the suggestion list is shown
+    showSuggestions: false,
+    // What the user has entered
+    userInput: ""
+}
+
+const propTypes = {
+	suggestions: PropTypes.array.isRequired,
+	suggestionsLabelField: PropTypes.string.isRequired,
+	suggestionsSidenoteField: PropTypes.string.isRequired,
+	name: PropTypes.string.isRequired,
+	placeholder: PropTypes.string.isRequired,
+	handleSelect: PropTypes.func.isRequired,
+	itemNamePlural: PropTypes.string.isRequired,
+}
+
+const filterSuggestions = query => ({
+	type: 'FILTER_SUGGESTIONS',
+	payload: query,
+})
+
+
+const selectSuggestion = query => ({
+	type: 'SELECT_SUGGESTION',
+	payload: query,
+})
+
+
+const incrementActiveSuggestion = {
+	type: 'INCREMENT_ACTIVE_SUGGESTION',
+	payload: null,
+}
+
+
+const decrementActiveSuggestion = {
+	type: 'DECREMENT_ACTIVE_SUGGESTION',
+	payload: null,
+}
+
+
+const AutoComplete = ({
+	suggestions,
+	suggestionsLabelField,
+	suggestionsSidenoteField,
+	name,
+	placeholder,
+	handleSelect,
+	itemNameSingular,
+	itemNamePlural,
+}) => {
+	const reducer = (state, {type, payload}) => {
+		switch (type) {
+			case 'SELECT_SUGGESTION':
+				return initialState
+			case 'DECREMENT_ACTIVE_SUGGESTION':
+				return {...state, activeSuggestion: Math.max(0, state.activeSuggestion - 1)}
+			case 'INCREMENT_ACTIVE_SUGGESTION':
+				return {...state, activeSuggestion: Math.min(state.filteredSuggestions.length - 1, state.activeSuggestion + 1)}
+			case 'FILTER_SUGGESTIONS':
+				const filteredSuggestions = suggestions.filter(
+					suggestion =>
+					suggestion[suggestionsLabelField].toLowerCase().indexOf(payload.toLowerCase()) > -1
+				)
+				return {
+					activeSuggestion: 0,
+					filteredSuggestions,
+					userInput: payload,
+					showSuggestions: true,
+				}
+
+			default:
+				throw new Error(`Unknown action type: ${type}`)
+		}
+	}
+
+	const [state, dispatch] = useReducer(reducer, initialState)
```

Comment:
TIL react has a built in `useReducer`—donno how that one went by without my notice!